File: wiz-spring-boilerplate/src/main/java/com/example/testcodesample/api/filter/JwtAuthenticationFilter.java
package com.example.testcodesample.api.filter;

import java.io.IOException;
import java.util.Optional;

import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import com.example.testcodesample.api.exception.ErrorCode;
import com.example.testcodesample.api.exception.UnAuthorizationException;
import com.example.testcodesample.common.JwtHelper;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

	private final JwtHelper jwtHelper;

	@Override
	protected void doFilterInternal(HttpServletRequest req, HttpServletResponse resp, FilterChain chain) throws ServletException, IOException {
		if (!this.isApplicable(req)) {
			chain.doFilter(req, resp);
			return;
		}
		String accessToken = Optional.ofNullable(req.getHeader("Authorization"))
			.map(header -> header.substring("Bearer ".length()))
			.orElseThrow(() -> new UnAuthorizationException(ErrorCode.NOT_FOUND_ACCESS_TOKEN));
		jwtHelper.validate(accessToken);
		chain.doFilter(req, resp);
	}

	public boolean isApplicable(HttpServletRequest req) {
		return req.getRequestURI().startsWith("/api");
	}
}

File: wiz-todo/src/main/kotlin/com/example/wiztodo/api/filter/JwtAuthenticationFilter.kt
package com.example.wiztodo.api.filter

import com.example.wiztodo.common.JwtHelper
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class JwtAuthenticationFilter(
    private val jwtHelper: JwtHelper
) : OncePerRequestFilter() {
    override fun doFilterInternal(
        req: HttpServletRequest,
        resp: HttpServletResponse,
        chain: FilterChain
    ) {
        /**
         * 1. 사용자로부터 인증에 필요한 AccessToken 을 받는다. 만약 없으면 인증 실패로 처리한다.
         * 2. AccessToken 이 정상적인지 검증한다.
         * (그다지 관심은 없지만.. 이 검증에는 토큰의 만료기간이 유효한지, 내가 만든 것인지 등을 검증한다)
         */
        val accessToken = req.getAccessToken() ?: throw RuntimeException()
        if (!jwtHelper.validate(accessToken)) {
            throw RuntimeException()
        }
        // if validate success
        chain.doFilter(req, resp)
    }

    fun HttpServletRequest.getAccessToken(): String? {
        return this.getHeader("Authorization") // Bearer {access_token}
            ?.substring("Bearer ".length)
    }
}